File: Brute Force/17472번 - 다리 만들기 2/17472.py
import sys

input = sys.stdin.readline

n, m = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(n)]
vis = [[False] * m for _ in range(n)]
edges = []
parents = list(range(n * m))


def dfs(x, y, num):
    global vis
    vis[x][y] = True
    arr[x][y] = num
    
    for dx, dy in [[0, -1], [-1, 0], [0, 1], [1, 0]]:
        nx = x + dx
        ny = y + dy
        if 0 <= nx < n and 0 <= ny < m:
            if not vis[nx][ny] and arr[nx][ny] == 1:
                dfs(nx, ny, num)

def bridge(x, y, land):
    for dx, dy in [[0, -1], [-1, 0], [0, 1], [1, 0]]:
        w = 0
        nx = x
        ny = y
        while(1):
            nx += dx
            ny += dy
            if nx < 0 or nx >= n or ny < 0 or ny >= m: 
                break
            new_island = arr[nx][ny]
            if new_island != 0:
                if new_island != land and w >= 2:
                    edges.append([w, land, new_island])
                break
            w += 1

def Find(x):
    if x != parents[x]:
        parents[x] = Find(parents[x])
    return parents[x]
    
def Union(a, b):
    a = Find(a)
    b = Find(b)
    if a != b:
        parents[b] = a

def solve():
    island_cnt = 1
    for i in range(n):
        for j in range(m):
            if arr[i][j] == 1 and not vis[i][j]:
                dfs(i, j, island_cnt)
                island_cnt += 1
    
    for i in range(n):
        for j in range(m):
            if arr[i][j] != 0:
                bridge(i, j, arr[i][j])
    
    edges.sort()
    edge_cnt = 0
    min_weight = 0
    chk = False
    for edge in edges:
        if Find(edge[1]) == Find(edge[2]):
            continue
        
        Union(edge[1], edge[2])
        min_weight += edge[0]
        edge_cnt += 1
        
        if edge_cnt == island_cnt - 2:
            chk = True
            break
    
    print(min_weight if chk else -1)


if __name__ == '__main__':
    solve()

File: Brute Force/17472번 - 다리 만들기 2/17472.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>

using namespace std;

#define X first
#define Y second

int n, m;
int island[11][11];
bool vis[11][11];
vector<pair<int, int> > v;
vector<vector<int> > bridge;
int dx[4] = {0, -1, 0, 1};
int dy[4] = {-1, 0, 1, 0};
int parents[101];

void input() {
    cin >> n >> m;
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
            cin >> island[i][j];
            if(island[i][j]) v.push_back({i, j});
        }
    }
    for(int i = 1; i <= 100; i++) {
        parents[i] = i;
    }
}

void connect(int x, int y) {
    int cur_island_num = island[x][y];
    
    for(int d = 0; d < 4; d++) {
        int o_x = x, o_y = y;
        int cnt = 0;
        int next_island = 0;
        while(1) {
            int nx = o_x + dx[d];
            int ny = o_y + dy[d];
            o_x = nx, o_y = ny;
            next_island = island[nx][ny];
            
            if(nx < 0 || nx >= n || ny < 0 || ny >= m || next_island == cur_island_num) break;
            if(next_island != 0) {
                if(cnt >= 2) {
                    bridge.push_back({cnt, island[x][y], island[nx][ny]});
                }
                break;
            }
            cnt += 1;
        }
    }
}

void bfs(int x, int y, int num) {
    queue<pair<int, int> > Q;
    Q.push({x, y});
    
    while(!Q.empty()) {
        pair<int, int> cur = Q.front(); Q.pop();
        island[cur.X][cur.Y] = num;
        vis[cur.X][cur.Y] = true;
        
        for(int i = 0; i < 4; i++) {
            int nx = cur.X + dx[i];
            int ny = cur.Y + dy[i];
            
            if(nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
            if(vis[nx][ny] || island[nx][ny] == 0) continue;
            vis[nx][ny] = true;
            Q.push({nx, ny});
        }
    }
}

int Find(int x) {
    if(x == parents[x]) return x;
    return parents[x] = Find(parents[x]);
}

void Union(int a, int b) {
    a = Find(a);
    b = Find(b);
    if(a != b) parents[b] = a;
}

void solve() {
    int island_num = 1;
    
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
            if(island[i][j] && !vis[i][j]) {
                bfs(i, j, island_num);
                island_num++;
            }
        }
    }
    
    for(auto &it : v) {
        connect(it.X, it.Y);
    }
    
    sort(bridge.begin(), bridge.end());
    
    int edge_cnt = 0;
    int min_weight = 0;
    bool chk = false;
    for(int i = 0; i < bridge.size(); i++) {
        vector<int> edge = bridge[i];

        if(Find(edge[1]) == Find(edge[2])) continue;
        Union(edge[1], edge[2]);
        min_weight += edge[0];

        if(++edge_cnt == island_num - 2) {
            chk = true;
            break;
        }
    }
    
    cout << (chk ? min_weight : -1) << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}